File: Tree/problem96/problem96.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int numTrees(int n) {
        // dp[i] = number of BST shapes with i nodes (values 1..i).
        // Root j has left subtree size j-1 and right subtree size i-j.
        vector<long long> dp(n + 1, 0);
        dp[0] = 1;
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= i; ++j) {
                dp[i] += dp[j - 1] * dp[i - j];
            }
        }
        return static_cast<int>(dp[n]);
    }
};

File: Tree/problem96/problem96.py
class Solution:
    def numTrees(self, n: int) -> int:
        # dp[i] = number of BST shapes with i nodes (values 1..i).
        # Root j has left subtree size j-1 and right subtree size i-j.
        dp = [0] * (n + 1)
        dp[0] = 1
        for i in range(1, n + 1):
            for j in range(1, i + 1):
                dp[i] += dp[j - 1] * dp[i - j]
        return dp[n]
